### PR TITLE
APP-driven catalog entity page with GitHub Actions integration & routes & tabs

### LINKED ISSUE #1859: GitHub Actions: provide a context with an entity to the plugin
A quote from Discord.

> Does the GitHub actions plugin work?
it's requesting this url:
https://api.github.com/repos///actions/runs?per_page=5&page=1
without the action repository
I hacked it with this:
```+++ b/plugins/github-actions/src/components/useProjectName.ts
@@ -23,7 +23,7 @@ export const useProjectName = (name: EntityCompoundName) => {

   const { value } = useAsync<string>(async () => {
     const entity = await catalogApi.getEntityByName(name);
-    return entity?.metadata.annotations?.['github.com/project-slug'] ?? '';
+    return entity?.metadata.annotations?.['github.com/project-slug'] ?? 'timja-org/rpe-sweet-backend';
   });
```
>I've set the annotation
but not sure how it's supposed to know which entity is in use
maybe a widget needs adding to the entity page
the docs just say to go to the github-actions page
but there's no context there

### LINKED ISSUE #1982: API for connecting plugin pages as tabs on entity pages

### PR DESCRIPTION
**Pull request by shmidt-i**
Draft PR that solves most of the #1536 discovered problems 💪
Work towards #1982 
in

## TLDR
- JSX-driven approach
- APP defines top-level routing in JSX (for the standlalone-pages types-of-plugins)
- Plugins define their own relative routing system using RouteRefs & ReactRouter v6 APIs
- Plugin-catalog exports smart component `CatalogPage` which accepts `EntityPage` as a prop
- `EntityPage` is defined inside the APP, using helpers from plugin-catalog (Tabs, useEntity, etc)
- Plugins, that supposed to live inside the `EntityPage` are now there in the APP, making it possible to compose/configure all you want out of it ;)

## TLDR (more details)

```tsx
// packages/app/App.tsx

import { CatalogPlugin } from '@backstage/plugin-catalog';
import { EntityPage } from './components/catalog';

const AppRoutes = () => (
  <Routes>
    <Route
      path="/catalog/*"
      element={<CatalogPlugin EntityPage={EntityPage} />}
    />
    {/* <Route path="/explore" element={<ExplorePlugin />} /> */}
  </Routes>
);
```

```tsx
// packages/app/components/catalog/index.tsx

import { useEntity } from '@backstage/plugin-catalog';
import { ComponentEntity } from './Component';

export const EntityPage = () => {
  const entity = useEntity();
  switch (entity.kind) {
    case 'Component':
      return <ComponentEntity />;
    default:
      return <UnkownEntityKind />;
  }
};
```

```tsx
// packages/app/components/catalog/Component.tsx

import { useEntity, EntityMetadataCard, EntityPageTabs as Tabs } from '@backstage/plugin-catalog';
import {GITHUB_ACTIONS_ANNOTATION, GithubActionsPlugin} from '@backstage/plugin-github-actions';

export const ComponentEntity = () => {
  const entity = useEntity();

  switch (entity.spec!.type) {
    case 'service':
      return <Service />;
    case 'website':
      return <Website />;
    default:
      return <UnkownEntity />;
  }
};

const Service = () => {
	const entity = useEntity();

  const isCIAvailable = [
    entity!.metadata!.annotations?.[GITHUB_ACTIONS_ANNOTATION],
    entity!.metadata!.annotations?.[CIRCLE_CI_ANNOTATION],
  ].some(Boolean);

  return (
    <Tabs>
      <Tabs.Tab title="Overview" path="/" exact>
        <OverviewPage />
      </Tabs.Tab>
      {isCIAvailable && (
        <Tabs.Tab title="CI/CD" path="/ci-cd">
          {entity!.metadata!.annotations?.[GITHUB_ACTIONS_ANNOTATION] && (
            <GitHubActionsPlugin entity={entity} />
          )}
        </Tabs.Tab>
      )}
      <Tabs.Tab title="Docs" path="/docs">
        <div>Docs tab contents</div>
      </Tabs.Tab>
    </Tabs>
  );}

const OverviewPage = () => {
  const entity = useEntity();

  return (
    <Grid item sm={4}>
      <EntityMetadataCard entity={entity} />
    </Grid>
  );
};
```

### CODE REVIEW

File: packages/app/src/components/catalog/Component.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  const entity = useEntity();
+
+  const isCIAvailable = [
+    entity!.metadata!.annotations?.[GITHUB_ACTIONS_ANNOTATION],
+    entity!.metadata!.annotations?.[CIRCLE_CI_ANNOTATION],
+  ].some(Boolean);
+
+  return (
+    <Tabs>
+      <Tabs.Tab title="Overview" path="/" exact>
```
Comment (freben) [inline packages/app/src/components/catalog/Component.tsx:50]: the exact thing can be deprecated under react-router 6 right?

File: packages/app/src/components/catalog/Component.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        <OverviewPage />
+      </Tabs.Tab>
+      {isCIAvailable && (
+        <Tabs.Tab title="CI/CD" path="/ci-cd">
+          {entity!.metadata!.annotations?.[GITHUB_ACTIONS_ANNOTATION] && (
+            <GitHubActionsPlugin entity={entity} />
+          )}
```
Comment (freben) [inline packages/app/src/components/catalog/Component.tsx:57]: forgot circleci here :) yeah i know it's illustrative

File: packages/app/src/components/catalog/Component.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  );
+};
+
+const Service = () => <DefaultEntity />;
+const Website = () => <DefaultEntity />;
+const UnkownEntity = () => <div>Unknown entity!</div>;
+
+export const ComponentEntity = () => {
+  const entity = useEntity();
+  switch (entity.spec!.type) {
```
Comment (freben) [inline packages/app/src/components/catalog/Component.tsx:74]: Question is if this switch is necessary - should we have to switch out / duplicate logic for choice of tabs, sidebar, etc per type? Maybe. Just wondering if it's the right "granularity".

**Review by freben — COMMENTED** [review]
sorry i only got partway ... will have to continue tomorrow

File: packages/app/src/components/catalog/Component.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        <OverviewPage />
+      </Tabs.Tab>
+      {isCIAvailable && (
+        <Tabs.Tab title="CI/CD" path="/ci-cd">
+          {entity!.metadata!.annotations?.[GITHUB_ACTIONS_ANNOTATION] && (
+            <GitHubActionsPlugin entity={entity} />
+          )}
+        </Tabs.Tab>
+      )}
+      <Tabs.Tab title="Docs" path="/docs">
```
Comment (emmaindal) [inline packages/app/src/components/catalog/Component.tsx:60]: I guess we want to have something similar to isCIAvailable here but isDocsAvailable or something and check the techdocs ref annotation to only show the Docs tab on entities that actually have docs?

Comment (shmidt-i) [reply packages/app/src/components/catalog/Component.tsx:60]: Yes, point being - since it's JSX, one can introduce whatever logic is necessary

Comment (emmaindal) [reply packages/app/src/components/catalog/Component.tsx:60]: ooor was that only for illustration purposes? xD

File: packages/app/src/components/catalog/Component.tsx (hunk 1 above)
Comment (shmidt-i) [reply packages/app/src/components/catalog/Component.tsx:50]: This is not the react-router `exact` though, it's part of the Tabs implementation

File: packages/app/src/components/catalog/Component.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+// Just for illustration purposes, gonna live in plugin-circle-ci
+const CIRCLE_CI_ANNOTATION = 'circle-ci-slug-dummy';
+
+const DefaultEntity = () => {
+  const entity = useEntity();
+
+  const isCIAvailable = [
+    entity!.metadata!.annotations?.[GITHUB_ACTIONS_ANNOTATION],
```
[the remark this answers is not in the document]
Comment (shmidt-i) [reply packages/app/src/components/catalog/Component.tsx:44]: Probably both things are fixable

File: packages/app/src/App.tsx [hunk 6]
```diff
@@ -46,8 +51,16 @@ const app = createApp({
 
 const AppProvider = app.getProvider();
 const AppRouter = app.getRouter();
-const AppRoutes = app.getRoutes();
 
+const AppRoutes = () => (
+  <Routes>
+    <Route
+      path="/catalog/*"
```
[the remark this answers is not in the document]
Comment (shmidt-i) [reply packages/app/src/App.tsx:58]: Not here, no, the * allows to have deeply nested subroutes

[the remark this answers is not in the document]
Comment (freben) [reply packages/app/src/App.tsx:58]: that doesn't seem to be the case here? https://reacttraining.com/blog/react-router-v6-pre/#nested-routes-and-layouts

File: packages/app/src/components/catalog/Component.tsx (hunk 1 above)
Comment (freben) [reply packages/app/src/components/catalog/Component.tsx:50]: Sure but maybe we want to change the tab implementation.

File: packages/app/src/App.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (shmidt-i) [reply packages/app/src/App.tsx:58]: https://reacttraining.com/blog/react-router-v6-pre/#introducing-routes

Same article, previous paragraph

File: plugins/catalog/src/routes.ts [hunk 7]
```diff
@@ -20,16 +20,16 @@ const NoIcon = () => null;
...
 export const entityRoute = createRouteRef({
   icon: NoIcon,
-  path: '/catalog/:kind/:optionalNamespaceAndName/:selectedTabId/*',
+  path: ':kind/:optionalNamespaceAndName/*',
   title: 'Entity',
 });
 export const entityRouteDefault = createRouteRef({
   icon: NoIcon,
-  path: '/catalog/:kind/:optionalNamespaceAndName',
+  path: ':kind/:optionalNamespaceAndName',
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply plugins/catalog/src/routes.ts:33]: The primary use-case for route-refs is linking into a plugin from other plugins or the app, doesn't this break that?

Btw, once we've settled the way we wanna do composition we want to have another look at route refs. Maybe they're not needed at all, but if they are we should convert the `path` to a method that takes params. In this case the type of the route ref would be something like `RouteRef<[kind: string, optionalNamespaceAndName: string]>`

File: packages/app/src/components/Root/Root.tsx [hunk 8]
```diff
@@ -89,7 +89,7 @@ const Root: FC<{}> = ({ children }) => (
       <SidebarSearchField onSearch={handleSearch} />
       <SidebarDivider />
       {/* Global nav, not org-specific */}
-      <SidebarItem icon={HomeIcon} to="./" text="Home" />
+      <SidebarItem icon={HomeIcon} to="/catalog" text="Home" />
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/app/src/components/Root/Root.tsx:92]: Yup a `routeRef` (`ref` is reserved 😅) prop for the `SidebarItem` is missing, we definitely want that.

We should preferably be able to generate a sidebar given a static config similar to this:

```yaml
sidebar:
  items:
  - ref: catalog.catalog
  - ref: scaffolder.create
  - divider
  - ref: explore.explore
  - ref: techdocs.docs
```

File: packages/app/src/components/catalog/Component.tsx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  ].some(Boolean);
+
+  return (
+    <Tabs>
+      <Tabs.Tab title="Overview" path="/" exact>
+        <OverviewPage />
+      </Tabs.Tab>
+      {isCIAvailable && (
+        <Tabs.Tab title="CI/CD" path="/ci-cd">
+          {entity!.metadata!.annotations?.[GITHUB_ACTIONS_ANNOTATION] && (
```
Comment (Rugvip) [inline packages/app/src/components/catalog/Component.tsx:55]: It will be tricky to generate an app based on static config if this kind of logic lives in the app. No need to fix as a part of this PR, but something we should look at after.

File: plugins/catalog/src/Router.tsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+        open={confirmationDialogOpen}
+        entity={entity!}
+        onConfirm={cleanUpAfterRemoval}
+        onClose={() => setConfirmationDialogOpen(false)}
+      />
+    </Page>
+  );
+};
+
+export const CatalogPlugin = ({ EntityPage }) => (
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply plugins/catalog/src/Router.tsx:142]: Yuuuuuuup. In general I think we can come up with a couple of suffixes for people to use depending on what is being exported. For example `*Page` and `*Card`.

I'm also thinking that all of these exported components will be wrapped by the plugin, so that we can inject a plugin context, pick up APIs, and ensure that they're all dynamically loaded.

File: plugins/catalog/src/routes.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Rugvip) [reply plugins/catalog/src/routes.ts:33]: The original RFC briefly hinted at this, but what I think we're really looking for here are immutable relative route refs. Something that lets you specify a route ref in relation to another route ref, where overrides on the second one would be reflected in the first.

File: packages/app/src/components/catalog/Component.tsx (hunk 2 above)
Comment (nikek) [reply packages/app/src/components/catalog/Component.tsx:57]: I think that the `GitHubActionsPlugin` component should not be named `Plugin`, as a plugin is the whole package, this component is the GithubActionsPage rather, IIUC.

Comment (nikek) [reply packages/app/src/components/catalog/Component.tsx:57]: Or is this the actual way to hook in a full plugin?

File: plugins/catalog/src/hooks/useEntity.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+      loading: null,
+      error: new Error('No name in url'),
+    } as never;
+  }
+
+  return { entity, loading, error };
+};
+
+/**
+ * Always going to return an entity, or throw an error if not a descendant of a EntityProvider.
```
Comment (shmidt-i) [inline plugins/catalog/src/hooks/useEntity.ts:71]: @freben - that's what you wanted, right?

Comment (freben) [reply plugins/catalog/src/hooks/useEntity.ts:71]: Yep. But I'm not sure about the type in there, looks complex :)

File: docs/plugins/index.md [hunk 12]
```diff
@@ -26,3 +26,10 @@ This helps the community know what plugins are in development.
 
 You can also use this process if you have an idea for a good plugin but you hope
 that someone else will pick up the work.
+
+## Integrate into the catalog service
```
Comment (stefanalund) [inline docs/plugins/index.md:30]: Service Catalog?

File: packages/app/src/components/catalog/EntityPage.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    <EntityPageLayout.Content
+      path="/ci-cd/*"
+      title="CI/CD"
+      element={<GitHubActionsRouter entity={entity} />}
+    />
+  </EntityPageLayout>
+);
+
+const DefaultEntityPage = ({ entity }: { entity: Entity }) => (
+  <EntityPageLayout>
```
Comment (shmidt-i) [inline packages/app/src/components/catalog/EntityPage.tsx:60]: Try to move on top of EntityPage

File: plugins/catalog/src/components/EntityPageLayout/Tabbed/Tabbed.tsx [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+ */
+export const Tabbed = {
+  Layout: ({ children }: { children: React.ReactNode }) => {
+    const routes: PartialRouteObject[] = [];
+    const tabs: Tab[] = [];
+    const params = useParams();
+    const navigate = useNavigate();
+
+    React.Children.forEach(children, child => {
+      if (!React.isValidElement(child)) {
```
Comment (shmidt-i) [inline plugins/catalog/src/components/EntityPageLayout/Tabbed/Tabbed.tsx:66]: Check if child type is actually `Content`

File: plugins/catalog/src/components/EntityPageLayout/Tabbed/Tabbed.tsx [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+
+    return (
+      <>
+        <HeaderTabs
+          tabs={tabs}
+          selectedIndex={selectedIndex}
+          onChange={onTabChange}
+        />
+        <Content>
+          <Grid container spacing={3}>
```
Comment (shmidt-i) [inline plugins/catalog/src/components/EntityPageLayout/Tabbed/Tabbed.tsx:114]: Remove `Grid`

File: plugins/catalog/src/hooks/useEntity.ts [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    if (error || (!loading && !entity)) {
+      errorApi.post(new Error('Entity not found!'));
+      setTimeout(() => {
+        navigate('/');
+      }, REDIRECT_DELAY);
+    }
+  }, [errorApi, navigate, error, loading, entity]);
+
+  if (!name) {
+    navigate('/catalog');
```
Comment (shmidt-i) [inline plugins/catalog/src/hooks/useEntity.ts:59]: just `/`

File: plugins/catalog/src/hooks/useEntity.ts [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+import { useApi, errorApiRef } from '@backstage/core';
+import { catalogApiRef } from '../api/types';
+import { useAsync } from 'react-use';
+import { Entity } from '@backstage/catalog-model';
+
+const REDIRECT_DELAY = 2000;
+
+type EntityLoadingStatus = {
+  entity?: Entity;
+  loading: boolean | null;
```
Comment (shmidt-i) [inline plugins/catalog/src/hooks/useEntity.ts:27]: boolean

File: plugins/catalog/src/hooks/useEntity.ts [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  error?: Error;
+};
+
+export const EntityContext = createContext<EntityLoadingStatus>({
+  entity: undefined as any,
+  loading: null,
```
Comment (shmidt-i) [inline plugins/catalog/src/hooks/useEntity.ts:33]: true in the beginning

File: plugins/catalog/src/routes.ts [hunk 19]
```diff
@@ -20,16 +20,16 @@ const NoIcon = () => null;
...
 export const entityRoute = createRouteRef({
   icon: NoIcon,
-  path: '/catalog/:kind/:optionalNamespaceAndName/:selectedTabId/*',
+  path: ':kind/:optionalNamespaceAndName/*',
   title: 'Entity',
 });
 export const entityRouteDefault = createRouteRef({
   icon: NoIcon,
-  path: '/catalog/:kind/:optionalNamespaceAndName',
+  path: ':kind/:optionalNamespaceAndName/*',
```
Comment (shmidt-i) [inline plugins/catalog/src/routes.ts:33]: Remove `/*`

**Comment by shmidt-i** [comment]
Leave old AppRoutes as well

**Review by stefanalund — APPROVED** [review]

**Review by Rugvip — APPROVED** [review]
